File: iosApp/iosApp/Features/Diary/notes/components/createnewnote/subnotes/SubNotesBlock.swift
//
//  SubNotesBlock.swift
//  iosApp
//
//  Created by velkonost on 26.10.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK

struct SubNotesBlock: View {
    
    let items: [SubNoteUI]
    @Binding var newSubNote: SubNoteUI
    
    let onNewSubNoteChanged: (String) -> Void
    let onAddNewSubNote: () -> Void
    let onSubNoteDelete: (SubNoteUI) -> Void
    
    @Binding private var isSubNotesBlockPickerVisible: Bool
    
    init(items: [SubNoteUI], newSubNote: Binding<SubNoteUI>, onNewSubNoteChanged: @escaping (String) -> Void, onAddNewSubNote: @escaping () -> Void, onSubNoteDelete: @escaping (SubNoteUI) -> Void, isSubNotesBlockPickerVisible: Binding<Bool>) {
        self.items = items
        self._newSubNote = newSubNote
        self.onNewSubNoteChanged = onNewSubNoteChanged
        self.onAddNewSubNote = onAddNewSubNote
        self.onSubNoteDelete = onSubNoteDelete
        self._isSubNotesBlockPickerVisible = isSubNotesBlockPickerVisible
    }
    
    var body: some View {
        PrimaryBox(
            padding: .init(top: .zero, leading: .zero, bottom: .zero, trailing: .zero)
        ) {
            VStack {
                SubNotesHeader(
                    isSubNotesBlockVisible: isSubNotesBlockPickerVisible,
                    onClick: {
                        withAnimation {
                            isSubNotesBlockPickerVisible.toggle()
                        }
                    }
                )
                
                if isSubNotesBlockPickerVisible {
                    LazyVStack {
                        ForEach(items, id: \.self.id) { item in
                            SubNoteItem(
                                item: item,
                                onDeleteSubNote: onSubNoteDelete
                            )
                        }
                        
                        AddSubNoteItem(
                            value: $newSubNote,
                            placeholderText: SharedR.strings().create_note_subnote_hint.desc().localized(),
                            onValueChanged: onNewSubNoteChanged,
                            onAddSubNote: onAddNewSubNote
                        )
                    }
                }
            }
        }
    }
}

File: androidApp/features/diary/src/main/kotlin/com/velkonost/getbetter/android/features/diary/notes/components/createnewnote/subnotes/SubNotesBlock.kt
package com.velkonost.getbetter.android.features.diary.notes.components.createnewnote.subnotes

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.runtime.MutableState
import androidx.compose.ui.Modifier
import com.velkonost.getbetter.core.compose.components.PrimaryBox
import com.velkonost.getbetter.shared.features.diary.model.SubNoteUI

@Composable
fun SubNotesBlock(
    modifier: Modifier = Modifier,
    items: List<SubNoteUI>,
    newSubNote: SubNoteUI,
    isSubNotesBlockVisible: MutableState<Boolean>,
    onNewSubNoteChanged: (String) -> Unit,
    onAddNewSubNote: () -> Unit,
    onSubNoteDelete: (SubNoteUI) -> Unit
) {
    PrimaryBox(padding = 0) {
        Column {
            SubNotesHeader(
                isSubNotesBlockVisible = isSubNotesBlockVisible
            )

            LazyColumn {
                items(items, key = { it.id }) { item ->
                    SubNoteItem(item = item)
                }

                item {
                    AddSubNoteItem(
                        value = newSubNote.text,
                        onValueChanged = onNewSubNoteChanged
                    )
                }
            }
        }
    }

}